### PR TITLE
OAuth flow fix

### LINKED ISSUE #1873: [UI Config] Fix Google OAuth flow

### PR DESCRIPTION
**Pull request by najclark**
## Description
These changes fix the oauth flow to be more consistent with the expected behavior seen on other websites. Previously, after signing in with an OAuth provider, the user would be redirected to the new user page, where they were prompted to create a password. 

Now when a user signs in with an OAuth provider, they are redirected to a page where they can see all of their courses.
Resolves #1873 

## How Has This Been Tested?
Follow the testing steps in #1870 to configure Google OAuth. 

### New user OAuth
1. Log out of Autolab
2. Click 'Sign in with Google' and select an account that **is not** currently registered on Autolab
3. Verify that the new user is redirected to the 'welcome to autolab' page

### Existing user OAuth
1. Log out of Autolab
2. Click 'Sign in with Google' and select an account that **is** currently registered on Autolab
3. Verify that the existing user is redirected to their courses page
4. Log out of Autolab again
5. Log in with the existing user's email and password to verify that standard auth still works

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -99,28 +99,60 @@ def self.find_for_shibboleth_oauth(auth, _signed_in_resource = nil)
...
-                                 uid: data["uid"])
       elsif (data = session["devise.shibboleth_data"])
         user.email = data["uid"] # email is uid in our case
-        user.authentications.new(provider: "CMU-Shibboleth",
-                                 uid: data["uid"])
+      else
+        return # early exit if not called from an oauth callback
       end
+
+      User.assign_random_password user
```
Comment (victorhuangwq) [inline app/models/user.rb:128]: So does this mean that a user is assigned a random password if they log in using oauth? does this happen once or every single time?

**Review by victorhuangwq — APPROVED** [review]
This is amazing.
Functionality wise it works well, but I am having difficulty understanding what exactly you have done to the code.

Could you explain to me two cases?:
1. What happens when a completely new user logs in, how is the account created etc?
2. What happens when I log in again?

At the same time, I'm going to tag @20wildmanj to have a second eye on this

**Comment by najclark** [comment]
1. What happens when a completely new user logs in, how is the account created etc?
When a completely new user logs in, a new autolab user is created with the details provided by the OAuth provider (email, first name, last name) and a random password. Additionally a new `Authentication` record is created to mark that this user has been authenticated with the OAuth provider.

2. What happens when I log in again?
When the completely new user logs in again with OAuth, `User.find_for_google_oauth2_oauth` finds the previously created `Authentication` record and the user is signed in.
If the completely new user wants to log in with standard auth, they will need to reset their password because they have no way of knowing the auto-generated password.

**Comment by victorhuangwq** [comment]
nit: could you make the text white on hover
image

**Comment by victorhuangwq** [comment]
So I tested out a funky user flow:

- I tried registering myself normally using my andrew email using the normal registration flow
- I don't verify myself from the email that was sent
- I then tried to log in using OAuth using my andrew email
- It tells me that I have to confirm my account
image

I'm wondering if a student who signed up normally, and then used OAuth, should just be allowed to skip the verification email?

**Comment by victorhuangwq** [comment]
Question:
Is using OAuth equivalent to checking the "remember me" button?

**Comment by najclark** [comment]
Will fix the flow you mentioned to skip confirmation when an oauth authentication method is added to a user.

Right now OAuth is equivalent to the standard sign in without _remember me_ checked (a session is generated). It's a pretty easy change to make it equivalent to _remember me_ (a session and remember me cookie is generated) but I wasn't sure what the standard practice with OAuth was.

File: app/models/user.rb [hunk 2]
```diff
@@ -99,28 +99,64 @@ def self.find_for_shibboleth_oauth(auth, _signed_in_resource = nil)
...
-                                 uid: data["uid"])
+        User.assign_random_password user
+        User.add_oauth_if_user_exists session
       end
     end
   end
 
+  def self.add_oauth_if_user_exists(session)
+    email, provider, uid = "", "", ""
+    if (data = session["devise.facebook_data"])
```
Comment (victorhuangwq) [inline app/models/user.rb:136]: are we still intending to do facebook in the future?

**Review by victorhuangwq — APPROVED** [review]
LGTM

**Comment by 20wildmanj** [comment]
Retroactively LGTM

**Pull request merged**